### DISCUSSION TITLE
How can I bind behaviors in Avalonia?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#14000] [Q&A]
I have different pages displayed as landscape and portrait screens. I want to share its behaviors in the same ViewModel.

For example:

ViewModel:
```
namespace AvaloniaApplication1
{
    public partial class ViewModel
    {       
        public void DragStarted(object? sender, Avalonia.Input.VectorEventArgs e) 
        { 
        }       
    }
}
```

XAML:
```
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"		
             xmlns:Local="using:AvaloniaApplication1"
             x:DataType="Local:ViewModel"	
             x:Class="AvaloniaApplication1">	
	<Grid>
		<Thumb DragStarted="{Binding DragStarted}"></Thumb>	
	</Grid>
</UserControl>
```

And it reports error that:

> Unable to find suitable setter or adder for property DragStarted of type Avalonia.Controls:Avalonia.Controls.Primitives.Thumb for argument Avalonia.Markup.Xaml:Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindingExtension, available setter parameter lists are:
> System.EventHandler`1<Avalonia.Input.VectorEventArgs>

What's wrong with my code and how can I bind it?

**Comment by thevortexcloud** [comment]
You can't bind to an event (although I think there is plans to allow that). That said MVVM does allow for the view to directly interact with the view model. As such you could just have the view pass what ever data it needs to the view model/call a method on it.

MVVM however does not allow for the view model to know about the view (or for the view model to directly call the view). As such it really should have no concept of things like touch and input. The view model should not care at all about presentation/how the user inputs things. It really sounds like instead you want to do this entirely in the view. In which case you could create a base view which both versions of your view inherit from.

**Reply by kyuranger** [reply]
I have another way by creating a base class that inherit from usercontrol and add all behaviors in it. When needed, just let the control inherit this base class.

Although this method is feasible, I think it is logically better to implement it directly on the ViewModel.

**Reply by thevortexcloud** [reply]
I suppose it really depends on how closely you want to follow MVVM. If you don't care then this is fine. But otherwise it's a violation. Ideally your view model should not even know Avalonia exists. That way all your main app logic is portable. But doing that often requires making a lot of abstractions and thinking carefully about your architecture. As such in this instance, a base user control class follows MVVM far better.

I personally would be trying my hardest to keep things like this out of my view models and instead deal with it in the view or pass the raw data to the view model if I had to. That way it avoids directly coupling the view model to Avalonia.

**Comment by kyuranger** [comment]
I found another way by using Avalonia XAML Behaviors(https://github.com/AvaloniaUI/Avalonia.Xaml.Behaviors)

And here are codes:
ViewModel:
```
namespace AvaloniaApplication1
{
    public partial class ViewModel
    {       
        public void DragStarted() 
        { 
        }       
    }
}
```

XAML:
```
<UserControl xmlns="https://github.com/avaloniaui"
  xmlns:i="clr-namespace:Avalonia.Xaml.Interactivity;assembly=Avalonia.Xaml.Interactivity"
 xmlns:ia="clr-namespace:Avalonia.Xaml.Interactions.Core;assembly=Avalonia.Xaml.Interactions"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"		
             xmlns:Local="using:AvaloniaApplication1"
             x:DataType="Local:ViewModel"	
             x:Class="AvaloniaApplication1">	
	<Grid>
		<Thumb Name="SlideDownThumb">
<i:Interaction.Behaviors>
				<ia:EventTriggerBehavior EventName="DragStarted" SourceObject="SlideDownThumb">
					<ia:InvokeCommandAction Command="{Binding DragStarted}" />
				</ia:EventTriggerBehavior>
</i:Interaction.Behaviors>
                </Thumb>	
	</Grid>
</UserControl>
```

However, this way seems to always need to set a name to the control.

What's more, this method can only realize part of the functions, such as
```
  private void DragDelta(object? sender, Avalonia.Input.VectorEventArgs e)
  {
   
  }
```

It throw error and can not get the `e.Vector.Y`.

**Reply by AWAS666** [reply]
yeah this also took me a while to figure out, but your post got me onto the right track.
you gotta set "PassEventArgsToCommand" to true, then it will pass the eventargs
This is false by default appearently...

**Reply by stevemonaco** [reply]
Default for good reason: by passing the event args to your ViewModel, you're breaking MVVM. Your ViewModel is referencing Avalonia-specific View types which it shouldn't know anything about.

**Reply by AWAS666** [reply]
how else do I bind the events if overwriting the datacontext with the view model makes binding to the view impossible?
or is there something I'm missing?

**Reply by stevemonaco** [reply]
Don't bind to events, events are meant to be handled in code-behind. IMO, the `EventTriggerBehavior` + `InvokeCommandAction` adapter is a bad pattern that has too much XAML ceremony. It's also not currently safe for trimming: https://github.com/AvaloniaUI/Avalonia.Xaml.Behaviors/blob/d62d2aa68e3943b26b8f33cc120f2e8cc707e038/src/Avalonia.Xaml.Interactions/Core/EventTriggerBehavior.cs#L63-L64

If an event needs to trigger work in the ViewModel, then my approach is:
1. Subscribe to the event handler and create an event handler in code-behind.
2. Adapt the args type to something ViewModel-neutral, if necessary.
3. Forward the call into the ViewModel by calling a method or invoking a command on it. All Views have access to their ViewModel via the `DataContext`.

You can avoid behaviors and sprinkle in any required View-specific functionality pre-or-post ViewModel invocation since you're in code.

**Comment by kyuranger** [comment]
There is another way here, which simply achieves the same requirement without using behavior binding.

Create a baseclass:
```
using Avalonia.Controls;
using Avalonia.Controls.Primitives;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AvaloniaApplication1
{
    public class BaseClass:UserControl
    {
        Thumb _thumnb;
        public BaseClass() {
            this.Loaded += BaseClass_Loaded;
        }

        private void BaseClass_Loaded(object? sender, Avalonia.Interactivity.RoutedEventArgs e)
        {
            if (this.FindControl<Thumb>("DragThumb") is Thumb i)
            { 
                _thumnb = i;
                _thumnb.DragDelta += _thumnb_DragDelta;
            }
        }

        private void _thumnb_DragDelta(object? sender, Avalonia.Input.VectorEventArgs e)
        {
            Debug.WriteLine(e.Vector.Y.ToString());
        }
    }
}
```

Create a usercontrol inherit baseclass:
```
using Avalonia;
using Avalonia.Controls;
using Avalonia.Markup.Xaml;

namespace AvaloniaApplication1;

public partial class UserControl1 : BaseClass
{
    public UserControl1()
    {
        InitializeComponent();
    }
}
```

The XAML of the usercontrol:
```
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"			 
             x:Class="AvaloniaApplication1.UserControl1">
	<Thumb Name="DragThumb" Height="200" Width="200" Background="Red">
		<Thumb.Styles>
			<Style Selector="Thumb">
				<Setter Property="Template">
					<ControlTemplate>
						<Border Background="{TemplateBinding Background}">							
						</Border>
					</ControlTemplate>
				</Setter>
			</Style>
		</Thumb.Styles>
	</Thumb>
</UserControl>
```

Finally, the MainWindow:
```
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
		xmlns:Local="using:AvaloniaApplication1"
        x:Class="AvaloniaApplication1.MainWindow"
        Title="AvaloniaApplication1">

	<Grid>
		<Local:UserControl1></Local:UserControl1>
	</Grid>
</Window>
```
